cplugin/server/serve.go

```diff
@@ -0,0 +1,157 @@
+// Copyright © 2024 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"context"
+	"errors"
+
+	"github.com/conduitio/conduit-connector-protocol/cplugin"
+	v1 "github.com/conduitio/conduit-connector-protocol/cplugin/v1"              //nolint:staticcheck // v1 is used for backwards compatibility
+	serverv1 "github.com/conduitio/conduit-connector-protocol/cplugin/v1/server" //nolint:staticcheck // v1 is used for backwards compatibility
+	v2 "github.com/conduitio/conduit-connector-protocol/cplugin/v2"
+	serverv2 "github.com/conduitio/conduit-connector-protocol/cplugin/v2/server"
+	connectorv1 "github.com/conduitio/conduit-connector-protocol/proto/connector/v1"
+	connectorv2 "github.com/conduitio/conduit-connector-protocol/proto/connector/v2"
+	"github.com/hashicorp/go-plugin"
+	"google.golang.org/grpc"
+)
+
+// Serve starts a go-plugin server with the given factories for creating
+// Specifier, Source, and Destination plugins. The server will support both v1
+// and v2 of the connector protocol.
+func Serve(
+	specifierFactory func() cplugin.SpecifierPlugin,
+	sourceFactory func() cplugin.SourcePlugin,
+	destinationFactory func() cplugin.DestinationPlugin,
+	opts ...Option,
+) error {
+	serveConfig := &plugin.ServeConfig{
+		HandshakeConfig: cplugin.HandshakeConfig,
+		VersionedPlugins: map[int]plugin.PluginSet{
+			v1.Version: v1PluginSetFactory(specifierFactory, sourceFactory, destinationFactory),
+			v2.Version: v2PluginSetFactory(specifierFactory, sourceFactory, destinationFactory),
```

Comment:
Notice that `Serve` will negotiate the version with Conduit, it accepts both v1 and v2.